File: backend/schemas/analytics.py
"""API contracts for Analytics Dashboard."""

from datetime import date, datetime
from pydantic import BaseModel, ConfigDict
from typing import Literal

class PerformanceStats(BaseModel):
    portfolio_value: float
    invested_amount: float
    absolute_return: float
    return_percentage: float | None

class RiskScoreComponent(BaseModel):
    score: int
    label: str

class RiskAssessment(BaseModel):
    overall_score: int | None
    label: str
    components: dict[str, RiskScoreComponent]
    explanations: list[str]

class ConcentrationMetrics(BaseModel):
    largest_position: float | None
    top_3: float | None
    hhi: float | None
    
class SnapshotDataPoint(BaseModel):
    date: str
    portfolio_value: float

class RecentDecision(BaseModel):
    date: str
    action: str
    asset: str
    weight_change: float
    reason: str

class AdaptiveThresholdState(BaseModel):
    previous: float
    observed: float
    updated: float
    change: float

class DataStats(BaseModel):
    portfolio_valuations: int
    holdings: int
    transactions: int
    rebalance_evaluations: int
    history_days: int
    last_updated: datetime | None

class AnalyticsDashboardResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    
    portfolio_id: str
    performance: PerformanceStats
    risk_assessment: RiskAssessment | None
    concentration: ConcentrationMetrics
    volatility: float | None
    max_drawdown: float | None
    performance_history: list[SnapshotDataPoint]
    recent_decisions: list[RecentDecision]
    adaptive_threshold: AdaptiveThresholdState | None
    data_stats: DataStats

File: frontend/src/api/analytics.ts
import { apiFetch } from "./client"

export interface PerformanceStats {
  portfolio_value: number
  invested_amount: number
  absolute_return: number
  return_percentage: number | null
}

export interface RiskScoreComponent {
  score: number
  label: string
}

export interface RiskAssessment {
  overall_score: number | null
  label: string
  components: Record<string, RiskScoreComponent>
  explanations: string[]
}

export interface ConcentrationMetrics {
  largest_position: number | null
  top_3: number | null
  hhi: number | null
}

export interface SnapshotDataPoint {
  date: string
  portfolio_value: number
}

export interface RecentDecision {
  date: string
  action: string
  asset: string
  weight_change: number
  reason: string
}

export interface AdaptiveThresholdState {
  previous: number
  observed: number
  updated: number
  change: number
}

export interface DataStats {
  portfolio_valuations: number
  holdings: number
  transactions: number
  rebalance_evaluations: number
  history_days: number
  last_updated: string | null
}

export interface AnalyticsDashboardResponse {
  portfolio_id: string
  performance: PerformanceStats
  risk_assessment: RiskAssessment | null
  concentration: ConcentrationMetrics
  volatility: number | null
  max_drawdown: number | null
  performance_history: SnapshotDataPoint[]
  recent_decisions: RecentDecision[]
  adaptive_threshold: AdaptiveThresholdState | null
  data_stats: DataStats
}

export const analyticsApi = {
  getDashboard: (portfolioId: string) =>
    apiFetch<AnalyticsDashboardResponse>(`/portfolios/${portfolioId}/analytics/dashboard`),
}
